rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/topic_cache.hpp

```diff
@@ -129,7 +166,9 @@ class TopicCache
         "Adding topic '%s' with type '%s' for node '%s'",
         topic_name.c_str(), type_name.c_str(), guid_stream.str().c_str());
     }
-    topic_to_types_[topic_name].push_back(type_name);
+    auto qos_profile = rmw_qos_profile_t();
+    dds_qos_to_rmw_qos(dds_qos, &qos_profile);
+    topic_name_to_topic_data_[topic_name].push_back(std::make_tuple(guid, type_name, qos_profile));
```

Comment:
You're storing the participant guid, and not the publisher guid.
The publisher/subscriber guid is also part of the discovery data, and it's the one that should be stored (you will have to add that argument to the function).

Reply:
Ah! So use [this](https://github.com/eProsima/Fast-RTPS/blob/master/include/fastdds/rtps/builtin/data/ReaderProxyData.h#L72), i.e use `proxyData.guid()` instead of `proxyData.RTPSParticipantKey()`. 
So in this case, `participant_to_topics_` will store the participant `guid` and the `topic_name_to_topic_data_` will store the `reader/writer` `guid`. 

``` c++
 bool addTopic(
    const eprosima::fastrtps::rtps::InstanceHandle_t & rtpsParticipantKey,
    const eprosima::fastrtps::rtps::GUID_t & guid /* publisher or subscriber GUID */
    const std::string & topic_name,
    const std::string & type_name)
    const T & dds_qos)
```

Is that correct?

However, now I have a question on how we can get the publisher/subscriber name and namespace; since `discovered_names` and `discovered_namespaces` is what I was using earlier and I am assuming that is what I need to use now as well. In which case I will need to have information about which participant a publisher/subscriber is associated with.

Reply:
Well, you will actually need both, the publisher/subscriber guid and the participant guid.
It would be better to create an struct instead of using a tuple, as if not it will be confusing which guid is what.